earthfile2llb/wait_block.go

```diff
@@ -0,0 +1,190 @@
+			}
+			if !hasPlatform {
+				return fmt.Errorf("SAVE IMAGE %s was already declared (none had --platform values)", saveImage.si.DockerTag)
+			}
+			if _, found := noManifestListImgs[saveImage.si.DockerTag]; found {
+				return fmt.Errorf("cannot save image %s defined multiple times, but declared as SAVE IMAGE --no-manifest-list", saveImage.si.DockerTag)
+			}
+		}
+
+		if saveImage.si.HasPlatform {
+			// SAVE IMAGE was called with a --platform value
+			if saveImage.si.NoManifestList {
+				noManifestListImgs[saveImage.si.DockerTag] = struct{}{}
+				isMultiPlatform[saveImage.si.DockerTag] = false
+			} else {
+				isMultiPlatform[saveImage.si.DockerTag] = true // do I need to count for previsouly seen?
+			}
+		} else {
+			isMultiPlatform[saveImage.si.DockerTag] = false
+		}
+		imageWaitItems = append(imageWaitItems, saveImage)
+	}
+	if len(imageWaitItems) == 0 {
+		return nil
+	}
+
+	metadata := map[string][]byte{}
+	refs := map[string]client.Reference{}
+
+	imageSaver := saveimageutil.NewExportImageSaver(refs, metadata)
+
+	refID := 0
+	for _, item := range imageWaitItems {
+		if !item.push {
+			continue
+		}
+
+		ref, err := llbutil.StateToRef(
+			ctx, item.c.opt.GwClient, item.si.State, item.c.opt.NoCache,
+			item.c.platr, item.c.opt.CacheImports.AsMap())
+		if err != nil {
+			return errors.Wrapf(err, "failed to solve image required for %s", item.si.DockerTag)
+		}
+
+		var platformBytes []byte
+		if isMultiPlatform[item.si.DockerTag] {
+			platformBytes = []byte(item.si.Platform.String())
+			platformImgName, err := llbutil.PlatformSpecificImageName(item.si.DockerTag, item.si.Platform)
+			if err != nil {
+				return err
+			}
+
+			if item.si.CheckDuplicate && item.si.DockerTag != "" {
+				if _, found := platformImgNames[platformImgName]; found {
+					return errors.Errorf(
+						"image %s is defined multiple times for the same platform (%s)",
+						item.si.DockerTag, platformImgName)
+				}
+				platformImgNames[platformImgName] = true
+			}
+		}
+
+		_, err = imageSaver.AddPushImageEntry(ref, refID, item.si.DockerTag, true, item.si.InsecurePush, item.si.Image, platformBytes)
+		if err != nil {
+			return err
+		}
+		refID++
+	}
+
+	if len(imageWaitItems) == 0 {
+		panic("saveImagesWaitItem should never have been created with zero converters")
+	}
+	gatewayClient := imageWaitItems[0].c.opt.GwClient // could be any converter's gwClient (they should app be the same)
+
+	err := gatewayClient.Export(ctx, gwclient.ExportRequest{
+		Refs:     refs,
+		Metadata: metadata,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to SAVE IMAGE")
+	}
+	return nil
+}
+
+func (wb *waitBlock) waitCommands(ctx context.Context) error {
+	errGroup, ctx := serrgroup.WithContext(ctx)
+
+	for _, item := range wb.items {
+		cmdItem, ok := item.(*runCommandWaitItem)
+		if !ok {
+			continue
+		}
+		errGroup.Go(func() error {
+			return cmdItem.c.forceExecution(ctx, *cmdItem.cmd, cmdItem.c.platr)
+		})
```

Comment:
We should limit parallelism based on the conversion parallelism semaphore perhaps?

If we do that, we should use a semaphore like this one: https://github.com/earthly/earthly/blob/main/earthfile2llb/withdockerrun.go#L72-L74. This ensures that at least one thread can make progress, since the assumption in the interpreter / converter parallelism code is that the main thread can always make progress.